rotary/rules/polish.smk

```diff
@@ -15,37 +15,120 @@ rule prepare_medaka_polish_input:
     input:
         "{sample}/assembly/{sample}_assembly.fasta"
     output:
-        temp("{sample}/polish/medaka_input/{sample}_input.fasta")
+        fasta=temp("{sample}/polish/medaka_input/{sample}_input.fasta"),
+        fasta_dir=temp(directory("{sample}/polish/medaka_input/"))
     run:
-        source_relpath = os.path.relpath(str(input),os.path.dirname(str(output)))
-        os.symlink(source_relpath,str(output))
-
+        source_relpath = os.path.relpath(str(input),os.path.dirname(str(output.fasta)))
+        os.symlink(source_relpath,str(output.fasta))
 
-rule polish_medaka:
+rule map_long_reads_for_medaka:
     input:
         qc_long_reads="{sample}/qc/{sample}_qc_long.fastq.gz",
         contigs="{sample}/{step}/medaka_input/{sample}_input.fasta"
     output:
-        dir=directory("{sample}/{step}/medaka"),
-        contigs="{sample}/{step}/medaka/{sample}_consensus.fasta",
-        calls_to_draft=temp(multiext("{sample}/{step}/medaka/calls_to_draft", '.bam', '.bam.bai')),
-        consensus_probs=temp("{sample}/{step}/medaka/consensus_probs.hdf")
+        calls_to_draft_bam=temp('{sample}/{step}/medaka/calls_to_draft.bam'),
+        calls_to_draft_bam_index=temp('{sample}/{step}/medaka/calls_to_draft.bam.bai')
     conda:
         "../envs/medaka.yaml"
     log:
-        "{sample}/logs/{step}/medaka.log"
+        "{sample}/logs/{step}/map_long_reads_medaka.log"
+    benchmark:
+        "{sample}/benchmarks/{step}/map_long_reads_medaka.txt"
+    threads:
+        config.get("threads",1)
+    params:
+        output_base_name='{sample}/{step}/medaka/calls_to_draft'
+    shell:
+        """
+        mini_align -i {input.qc_long_reads} -r {input.contigs} -m -p {params.output_base_name} -t {threads} > {log} 2>&1
+        """
+
+checkpoint generate_contig_manifest:
+    input:
+        "{sample}/{step}/medaka_input/{sample}_input.fasta"
+    output:
+        contig_manifest="{sample}/{step}/medaka/{sample}_contigs.txt"
+    conda:
+        "../envs/medaka.yaml"
+    log:
+        "{sample}/logs/{step}/contig_names_medaka.log"
     benchmark:
         "{sample}/benchmarks/{step}/medaka.txt"
+    shell:
+        """
+        seqkit seq -n {input} > {output.contig_manifest} 2>> {log}
```

Comment:
Yes, this dumps a list of contig identifiers to a file. 

Over the years, I have generally leaned towards using a dedicated bioinformatics tool rather than Unix tools to do header extraction because I have found that people are very liberal with the text of their faster headers. For example, people put `-->` in their fasta headers and other weird characters, or they put spaces or they put "(" and ")". All these things tend to break the parsers you write yourself. 

If you use a dedicated tool, the developer has coded in logic to strip out bad characters or other info, which makes things stabler.

Reply:
However, because this step will only get fasta files generated by an assembler as input we might be able to go to grep, awk, or sed. What are your thoughts?

Reply:
@jmtsuji Swapped out for shell commands.

Reply:
@LeeBergstrand Thanks for the discussion! Yeah, that makes sense to try to be careful, in general, of weird things in FastA headers. Because we are working more-or-less directly with the output from Flye like you mentioned, I think the shell code you've added should be OK. I think the shell code might need a minor modification so that STDERR from all three commands go to the log file:

Current
```bash
grep '^>' {input} | cut -f1 -d' ' | tr -d '>' > {output} 2>> {log}
```

Modified (with a subshell):
```bash
(grep '^>' {input} | cut -f1 -d' ' | tr -d '>' > {output}) 2> {log}
```
Note that I also changed `2>>` to `2>` so that it will overwrite rather than append to an existing log.

Does that look okay to you? If so, I am supportive of switching to this shell-based approach... it might leave us with one less software version to keep updated in the long term. Hopefully the shell code is simple enough to be compatible across multiple platforms and not easily broken... I tested the modified code in both bash and sh, and it seems to work as expected.

Reply:
@jmtsuji I will make the change you requested to the shell code. Yes, I think we should go with the shell code.